Source/Surfaces/BiCubicClass.f90

```diff
@@ -0,0 +1,429 @@
+         DO i = 4, 1, -1
+            z_val = z_val * v + ( ( coeffs(i,4) * u + coeffs(i,3) ) * u + coeffs(i,2) ) * u + coeffs(i,1)
+         END DO ! i
+
+      END FUNCTION evaluateBicubicInterpolant
+!
+!////////////////////////////////////////////////////////////////////////
+!
+      FUNCTION SparseCCS_MxV(x) RESULT(y)
+      ! Compressed column storage (CCS) version of matrix vector product y = Ax
+      ! The CCS version of this matrix is precomputed and stored above in the
+      ! variables 'values_weight_Matrix', 'row_index_weight_Matrix', and
+      ! 'column_pointers_weight_Matrix'.
+      IMPLICIT NONE
+      REAL(KIND=RP) :: x(16)
+      REAL(KIND=RP) :: y(16)
+      ! local variables
+      INTEGER :: j, k1, k2
+
+      y = 0.0_RP
+      DO j = 1,16
+         k1 = column_pointers_weight_Matrix(j)
+         k2 = column_pointers_weight_Matrix(j + 1) - 1
+         y(row_index_weight_Matrix(k1:k2)) = y(row_index_weight_Matrix(k1:k2)) &
+                                             + values_weight_Matrix(k1:k2) * x(j)
+      END DO ! j
+
+      END FUNCTION SparseCCS_MxV
+!
+!////////////////////////////////////////////////////////////////////////
+!
+      INTEGER FUNCTION getLeftIndex(values_list, test_point, N) RESULT(idx)
+         ! Naive loop through a set of data in values_list to determine the left index
+         ! of the interval that contains test_point
+         ! ACHTUNG! Assumes that the data in values_list is ordered
+         IMPLICIT NONE
+         INTEGER                    , INTENT(IN) :: N
+         REAL(KIND=RP)              , INTENT(IN) :: test_point
+         REAL(KIND=RP), DIMENSION(N), INTENT(IN) :: values_list
+         ! local variables
+         INTEGER :: j
+
+         DO j = 1,N-1
+            IF ( ( test_point .GE. values_list(j) ) .AND. ( test_point .LE. values_list(j+1) ) ) THEN
+               idx = j
+               EXIT
+            END IF
+         END DO ! j
+      END FUNCTION getLeftIndex
+!
+!////////////////////////////////////////////////////////////////////////
+!
+      FUNCTION pullFourCorners(z_array, x_idx, y_idx, nx, ny)  RESULT(corns)
+         ! Extract four corner values (in anti-clockwise orientation)
+         ! ACHTUNG 1! Assumes ordered grid data is provided by the file
+         ! ACHTUNG 2! Assumes the same number of points in each direction for the z values array
+         IMPLICIT NONE
+         INTEGER                         , INTENT(IN) :: nx, ny, x_idx, y_idx
+         REAL(KIND=RP), DIMENSION(nx, ny), INTENT(IN) :: z_array
+         REAL(KIND=RP)                                :: corns(4)
+
+         corns(1) = z_array( x_idx     , y_idx     )
+         corns(2) = z_array( x_idx + 1 , y_idx     )
+         corns(3) = z_array( x_idx + 1 , y_idx + 1 )
+         corns(4) = z_array( x_idx     , y_idx + 1 )
+
+      END FUNCTION pullFourCorners
+!
+!//////////////////////////////////////////////////////////////////////// 
+!
+!     TESTING FUNCTIONS FOLLOW
```

Comment:
Thanks for writing tests for this!

Reply:
I needed to do it anyway to make sure my code worked, and now I have a sort of workflow for something like this, with makes it easy to just add the testing to the main code.